instrumentation/netty-3.8/src/main/java/com/splunk/opentelemetry/netty/v3_8/NettyChannelPipelineInstrumentation.java

```diff
@@ -0,0 +1,146 @@
+      // don't end up with duplicates (this throws an exception)
+      if (pipeline.get(handler.getClass().getName()) != null) {
+        pipeline.remove(handler.getClass().getName());
+      }
+      // CallDepth does not allow just getting the depth value, so to avoid interfering with the
+      // upstream netty implementation we do the same count but with our class
+      return CallDepthThreadLocalMap.incrementCallDepth(ServerTimingHandler.class);
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void addHandler(
+        @Advice.Enter int depth,
+        @Advice.This ChannelPipeline pipeline,
+        @Advice.Argument(1) ChannelHandler handler) {
+      if (depth > 0) {
+        return;
+      }
+
+      ContextStore<Channel, ChannelTraceContext> contextStore =
+          InstrumentationContext.get(Channel.class, ChannelTraceContext.class);
+
+      try {
+        if (handler instanceof HttpServerCodec || handler instanceof HttpResponseEncoder) {
+          pipeline.addLast(
+              ServerTimingHandler.class.getName(), new ServerTimingHandler(contextStore));
+        }
+      } finally {
+        CallDepthThreadLocalMap.reset(ServerTimingHandler.class);
+      }
+    }
+  }
+
+  public static class ChannelPipelineAdd3ArgsAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static int checkDepth(
+        @Advice.This ChannelPipeline pipeline, @Advice.Argument(2) ChannelHandler handler) {
+      // Pipelines are created once as a factory and then copied multiple times using the same add
+      // methods as we are hooking. If our handler has already been added we need to remove it so we
+      // don't end up with duplicates (this throws an exception)
+      if (pipeline.get(handler.getClass().getName()) != null) {
+        pipeline.remove(handler.getClass().getName());
+      }
+      // CallDepth does not allow just getting the depth value, so to avoid interfering with the
+      // upstream netty implementation we do the same count but with our class
+      return CallDepthThreadLocalMap.incrementCallDepth(ServerTimingHandler.class);
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void addHandler(
+        @Advice.Enter int depth,
+        @Advice.This ChannelPipeline pipeline,
+        @Advice.Argument(2) ChannelHandler handler) {
+      if (depth > 0) {
+        return;
+      }
+
+      ContextStore<Channel, ChannelTraceContext> contextStore =
+          InstrumentationContext.get(Channel.class, ChannelTraceContext.class);
+
+      try {
+        if (handler instanceof HttpServerCodec || handler instanceof HttpResponseEncoder) {
+          pipeline.addLast(
+              ServerTimingHandler.class.getName(), new ServerTimingHandler(contextStore));
+        }
+      } finally {
+        CallDepthThreadLocalMap.reset(ServerTimingHandler.class);
+      }
+    }
+  }
```

Comment:
Any reason why implementation is duplicated? AFAIU the only difference is advices is the position of `ChannelHandler` argument – 1 vs 2. Other than that logic seems to be the same. Maybe extract that in the abstract super-class?

Reply:
Hmm, I can extract some common parts to an additional helper class - abstract class probably wouldn't work that well, since advice classes are not "normal" classes.